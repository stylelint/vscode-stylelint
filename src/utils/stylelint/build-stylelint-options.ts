import path from 'path';
import pathIsInside from 'path-is-inside';
import { URI } from 'vscode-uri';
import { findPackageRoot } from '../packages/index';
import type stylelint from 'stylelint';
import type { RunnerOptions } from './types';

/**
 * Given a document URI, base options, and extension options, builds a Stylelint
 * options object. Runner options supersede base options.
 */
export async function buildStylelintOptions(
	uri: string,
	workspaceFolder?: string,
	baseOptions: Partial<stylelint.LinterOptions> = {},
	{
		config,
		configFile,
		configBasedir,
		customSyntax,
		ignoreDisables,
		reportDescriptionlessDisables,
		reportNeedlessDisables,
		reportInvalidScopeDisables,
	}: RunnerOptions = {},
): Promise<Partial<stylelint.LinterOptions>> {
	const options = {
		...baseOptions,

		config: config ?? baseOptions.config,

		configFile: configFile
			? workspaceFolder
				? configFile.replace(/\$\{workspaceFolder\}/gu, workspaceFolder)
				: configFile
			: baseOptions.configFile,

		configBasedir: configBasedir
			? path.isAbsolute(configBasedir)
				? configBasedir
				: path.join(workspaceFolder ?? '', configBasedir)
			: baseOptions.configBasedir,

		customSyntax: customSyntax
			? workspaceFolder
				? customSyntax.replace(/\$\{workspaceFolder\}/gu, workspaceFolder)
				: customSyntax
			: baseOptions.customSyntax,

		ignoreDisables: ignoreDisables ?? baseOptions.ignoreDisables,

		reportDescriptionlessDisables:
			reportDescriptionlessDisables ?? baseOptions.reportDescriptionlessDisables,

		reportNeedlessDisables: reportNeedlessDisables ?? baseOptions.reportNeedlessDisables,

		reportInvalidScopeDisables:
			reportInvalidScopeDisables ?? baseOptions.reportInvalidScopeDisables,
	};

	const documentPath = URI.parse(uri).fsPath;

	if (documentPath) {
		if (workspaceFolder && pathIsInside(documentPath, workspaceFolder)) {
			options.ignorePath = path.join(workspaceFolder, '.stylelintignore');
		}

		if (options.ignorePath === undefined) {
			options.ignorePath = path.join(
				(await findPackageRoot(documentPath)) || path.parse(documentPath).root,
				'.stylelintignore',
			);
		}
	}

	return options;
}
