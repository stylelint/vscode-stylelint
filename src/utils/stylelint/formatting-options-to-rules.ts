import type LSP from 'vscode-languageserver-protocol';

export type FormattingRules = {
	indentation?: [number | string];
	'no-missing-end-of-source-newline'?: true | null;
	'no-eol-whitespace'?: true | null;
};

/**
 * Converts the given formatting options to rules.
 * @param options The formatting options.
 * @returns The rules.
 */
export function formattingOptionsToRules({
	insertSpaces,
	tabSize,
	insertFinalNewline,
	trimTrailingWhitespace,
}: LSP.FormattingOptions): FormattingRules {
	// NOTE: There is no equivalent rule for trimFinalNewlines, so it is not respected.
	// TODO: Create respective rule upstream?

	const rules: FormattingRules = {
		indentation: [insertSpaces ? tabSize : 'tab'],
	};

	if (insertFinalNewline !== undefined) {
		rules['no-missing-end-of-source-newline'] = insertFinalNewline || null;
	}

	if (trimTrailingWhitespace !== undefined) {
		rules['no-eol-whitespace'] = trimTrailingWhitespace || null;
	}

	return rules;
}
