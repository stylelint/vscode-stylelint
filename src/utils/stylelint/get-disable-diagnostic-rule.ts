import type LSP from 'vscode-languageserver-protocol';
import { DisableReportRuleNames } from './types';

/**
 * Gets the rule name to which a disable diagnostic applies. Returns `undefined`
 * if the diagnostic is not a disable diagnostic.
 * @param diagnostic The diagnostic corresponding to the Stylelint warning.
 */
export function getDisableDiagnosticRule(diagnostic: LSP.Diagnostic): string | undefined {
	// eslint-disable-next-line @typescript-eslint/switch-exhaustiveness-check
	switch (diagnostic.code) {
		case DisableReportRuleNames.Needless:
			return diagnostic.message.match(/^Needless disable for "(.+)"$/)?.[1];

		case DisableReportRuleNames.InvalidScope:
			return diagnostic.message.match(/^Rule "(.+)" isn't enabled$/)?.[1];

		case DisableReportRuleNames.Descriptionless:
			return diagnostic.message.match(/^Disable for "(.+)" is missing a description$/)?.[1];

		case DisableReportRuleNames.Illegal:
			return diagnostic.message.match(/^Rule "(.+)" may not be disabled$/)?.[1];

		default:
			return undefined;
	}
}
