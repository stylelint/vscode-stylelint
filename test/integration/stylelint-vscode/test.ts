import { join, resolve } from 'path';
import { pathToFileURL } from 'url';
import { TextDocument } from 'vscode-languageserver-textdocument';
import { StylelintRunner } from '../../../src/utils/stylelint/index';
import { version as stylelintVersion } from 'stylelint/package.json';
import { version as stylelintScssVersion } from 'stylelint-scss/package.json';
import semver from 'semver';

const createDocument = (uri: string | null, languageId: string, contents: string): TextDocument =>
	TextDocument.create(
		uri ? pathToFileURL(resolve(__dirname, '..', uri)).toString() : 'Untitled:Untitled',
		languageId,
		1,
		contents,
	);

describe('StylelintRunner', () => {
	test('should be resolved with diagnostics when it lints CSS successfully', async () => {
		expect.assertions(1);
		const runner = new StylelintRunner();
		const result = await runner.lintDocument(createDocument(null, 'css', '  foo { color: #y3 }'), {
			config: {
				rules: {
					'color-no-invalid-hex': [true],
					'selector-type-no-unknown': [true],
				},
			},
		});

		expect(result.diagnostics).toMatchSnapshot();
	});

	test('should be resolved with an empty array when no errors and warnings are reported', async () => {
		expect.assertions(1);
		const runner = new StylelintRunner();
		const result = await runner.lintDocument(createDocument(null, 'scss', ''), {
			config: {
				customSyntax: 'postcss-scss',
				rules: { 'color-no-invalid-hex': [true] },
			},
		});

		expect(result.diagnostics).toEqual([]);
	});

	test('should be resolved with one diagnostic when the CSS is broken', async () => {
		expect.assertions(1);
		const runner = new StylelintRunner();
		// TODO: Restore once postcss-markdown is PostCSS 8 compatible
		// 		const result = await runner.lintDocument(
		// 			createDocument(
		// 				'markdown.md',
		// 				'markdown',
		// 				`# Title

		// # Code block

		// \`\`\`css
		//           a{
		// \`\`\`
		// `,
		// 			),
		// 			{
		// 				config: {
		// 					customSyntax: 'postcss-markdown',
		// 					rules: {
		// 						indentation: ['tab'],
		// 					},
		// 				},
		// 			},
		// 		);
		const result = await runner.lintDocument(
			createDocument('scss.scss', 'scss', '          a{color: #y3\n'),
			{
				config: {
					customSyntax: 'postcss-scss',
					rules: {
						'color-no-invalid-hex': [true],
					},
				},
			},
		);

		expect(result.diagnostics).toMatchSnapshot();
	});

	test('should be resolved even if no configs are defined', async () => {
		expect.assertions(1);
		const runner = new StylelintRunner();
		// TODO: Restore once postcss-html is PostCSS 8 compatible
		// const result = await runner.lintDocument(createDocument(null, 'plaintext', '<style>a{}</style>'), {
		// 	customSyntax: 'postcss-html',
		// });
		const result = await runner.lintDocument(createDocument(null, 'plaintext', 'a{}'), {
			customSyntax: 'postcss-scss',
		});

		expect(result.diagnostics).toEqual([]);
	});

	test('should support `.stylelintignore`.', async () => {
		expect.assertions(1);
		const runner = new StylelintRunner();
		const result = await runner.lintDocument(
			createDocument('listed-in-stylelintignore.css', 'css', '}'),
			{
				ignorePath: require.resolve('./.stylelintignore'),
			},
		);

		expect(result.diagnostics).toEqual([]);
	});

	test('should support CSS-in-JS with customSyntax', async () => {
		expect.assertions(1);
		const runner = new StylelintRunner();
		const result = await runner.lintDocument(
			createDocument(
				null,
				'javascript',
				`import glamorous from 'glamorous';
const styled = require("styled-components");
const A = glamorous.a({font: 'bold'});
const B = styled.b\`
font: normal
\`;`,
			),
			{
				config: {
					customSyntax: '@stylelint/postcss-css-in-js',
					rules: { 'font-weight-notation': ['numeric'] },
				},
			},
		);

		expect(result.diagnostics).toMatchSnapshot();
	});

	test('should set `codeFilename` option from a TextDocument', async () => {
		expect.assertions(1);
		const runner = new StylelintRunner();
		const result = await runner.lintDocument(
			createDocument(
				'should-be-ignored.xml',
				'xml',
				`<style>
a { color: #000 }
</style>`,
			),
			{
				config: {
					rules: {
						'color-hex-length': 'long',
					},
					ignoreFiles: '**/*-ignored.xml',
				},
			},
		);

		expect(result.diagnostics).toEqual([]);
	});

	if (semver.satisfies(stylelintVersion, '^14')) {
		test('should support `processors` option', async () => {
			expect.assertions(1);
			const runner = new StylelintRunner();
			const result = await runner.lintDocument(
				createDocument('processors.tsx', 'typescriptreact', 'styled.p`"`'),
				{
					config: {
						// eslint-disable-next-line @typescript-eslint/ban-ts-comment -- for stylelint v14
						// @ts-ignore for stylelint v14
						processors: ['stylelint-processor-styled-components'],
						rules: {},
					},
				},
			);

			expect(result.diagnostics).toEqual([
				{
					code: 'CssSyntaxError',
					message: 'Unclosed string (CssSyntaxError)',
					range: {
						end: {
							character: 10,
							line: 0,
						},
						start: {
							character: 9,
							line: 0,
						},
					},
					severity: 1,
					source: 'Stylelint',
				},
			]);
		});
	}

	test('should check CSS syntax even if no configuration is provided', async () => {
		expect.assertions(1);
		const runner = new StylelintRunner();
		const result = await runner.lintDocument(
			createDocument('unclosed.css', 'css', 'a{color:rgba(}'),
		);

		expect(result.diagnostics).toMatchSnapshot();
	});

	test('should check CSS syntax even if no rule is provided', async () => {
		expect.assertions(1);
		const runner = new StylelintRunner();
		// TODO: Restore once postcss-html is PostCSS 8 compatible
		// const result = await runner.lintDocument(createDocument('at.xsl', 'xsl', '<style>@</style>'), {
		// 	customSyntax: 'postcss-html',
		// });
		const result = await runner.lintDocument(createDocument('at.scss', 'scss', '@'), {
			customSyntax: 'postcss-scss',
		});

		expect(result.diagnostics).toMatchSnapshot();
	});

	test('should be resolved even if no rules are defined', async () => {
		expect.assertions(1);
		const runner = new StylelintRunner();
		const result = await runner.lintDocument(createDocument('no-rules.css', 'css', 'a{}'), {
			config: {},
		});

		expect(result.diagnostics).toEqual([]);
	});

	test('should reject with a reason when it takes incorrect options', async () => {
		expect.assertions(1);
		const runner = new StylelintRunner();
		const promise = runner.lintDocument(
			createDocument('invalid-options.css', 'css', '  foo { color: #y3 }'),
			{
				config: {
					rules: {
						'color-no-invalid-hex': true,
						'color-hex-alpha': 'foo',
						'selector-type-no-unknown': [true, { bar: true }],
					},
				},
			},
		);

		await expect(promise).rejects.toThrowErrorMatchingSnapshot();
	});

	test('should be resolved with diagnostics when the rules include unknown rules', async () => {
		expect.assertions(1);
		const runner = new StylelintRunner();
		const result = await runner.lintDocument(createDocument('unknown-rule.css', 'css', 'b{}'), {
			config: {
				rules: {
					'this-rule-does-not-exist': 1,
					'this-rule-also-does-not-exist': 1,
				},
			},
		});

		expect(result.diagnostics).toMatchSnapshot();
	});

	if (semver.satisfies(stylelintScssVersion, '^15')) {
		test('should be resolved with diagnostic plugin rule URL', async () => {
			expect.assertions(1);
			const runner = new StylelintRunner();
			const result = await runner.lintDocument(
				createDocument('unknown-rule.scss', 'scss', '@unknown (max-width: 960px) {}'),
				{
					config: {
						plugins: ['stylelint-scss'],
						rules: {
							'scss/at-rule-no-unknown': true,
						},
					},
				},
			);

			expect(result.diagnostics).toEqual([
				{
					code: 'scss/at-rule-no-unknown',
					codeDescription: {
						href: 'https://github.com/stylelint-scss/stylelint-scss/blob/master/src/rules/at-rule-no-unknown',
					},
					message: 'Unexpected unknown at-rule "@unknown" (scss/at-rule-no-unknown)',
					range: {
						end: {
							character: 8,
							line: 0,
						},
						start: {
							character: 0,
							line: 0,
						},
					},
					severity: 1,
					source: 'Stylelint',
				},
			]);
		});
	}
});

describe('StylelintRunner with a configuration file', () => {
	test('should adhere to configuration file settings', async () => {
		expect.assertions(1);
		const runner = new StylelintRunner();
		const result = await runner.lintDocument(
			createDocument(
				join(__dirname, 'has-config-file.tsx'),
				'typescriptreact',
				`
const what: string = "is this";
<a css={{
  width: "0px",
  what
}} />;
`,
			),
			{ configFile: join(__dirname, 'no-unknown.config.js') },
		);

		expect(result.diagnostics).toMatchSnapshot();
	});
});

describe('StylelintRunner with auto-fix', () => {
	test('auto-fix should work properly if configs are defined', async () => {
		expect.assertions(1);
		const runner = new StylelintRunner();
		const result = await runner.lintDocument(
			createDocument(null, 'css', 'a\n{\ncolor:#ffffff;\n}'),
			{
				config: { rules: { 'color-hex-length': 'short' } },
				fix: true,
			},
		);

		expect(result.output).toMatchSnapshot();
	});

	test('auto-fix should only work properly for syntax errors if no rules are defined', async () => {
		expect.assertions(1);
		const runner = new StylelintRunner();
		const result = await runner.lintDocument(createDocument('no-rules.css', 'css', 'a {'), {
			config: {},
			fix: true,
		});

		expect(result.output).toBe('a {}');
	});

	test('JS file auto-fix should not change the content if no rules are defined', async () => {
		expect.assertions(1);
		const runner = new StylelintRunner();
		const result = await runner.lintDocument(createDocument('no-rules.js', 'javascript', '"a"'), {
			customSyntax: '@stylelint/postcss-css-in-js',
			config: {},
			fix: true,
		});

		expect(result.output).toBe('"a"');
	});

	test('auto-fix should ignore if the file matches the ignoreFiles', async () => {
		expect.assertions(1);
		const runner = new StylelintRunner();
		const result = await runner.lintDocument(
			createDocument('should-be-ignored.js', 'javascript', '"a"'),
			{
				customSyntax: '@stylelint/postcss-css-in-js',
				config: {
					rules: {},
					ignoreFiles: '**/*-ignored.js',
				},
				fix: true,
			},
		);

		expect(result.output).toBeUndefined();
	});

	test('auto-fix should work if there is syntax errors in css', async () => {
		expect.assertions(1);
		const runner = new StylelintRunner();
		const result = await runner.lintDocument(
			createDocument(
				'test.css',
				'css',
				`
.a {
    color: #ffffff
    background-color: #ffffffaa;
}
`,
			),
			{
				config: { rules: {} },
				fix: true,
			},
		);

		expect(result.output).toMatchSnapshot();
	});

	test('auto-fix should ignore if there is syntax errors in scss', async () => {
		expect.assertions(1);
		const runner = new StylelintRunner();
		const result = await runner.lintDocument(
			createDocument(
				'test.scss',
				'scss',
				`
.a {
    color: #ffffff
    background-color: #ffffffaa;
}
`,
			),
			{
				customSyntax: 'postcss-scss',
				config: { rules: {} },
				fix: true,
			},
		);

		expect(result.output).toBeUndefined();
	});

	test('auto-fix should work if there are errors that cannot be auto-fixed', async () => {
		const runner = new StylelintRunner();
		const result = await runner.lintDocument(
			createDocument(
				'test.css',
				'css',
				`
unknown {
    color: #ffffff;
    background-color: #ffffffaa;
}
`,
			),
			{
				config: {
					rules: {
						'color-hex-length': 'short',
						'selector-type-no-unknown': true,
					},
				},
				fix: true,
			},
		);

		expect(result).toMatchSnapshot();
	});
});

describe('StylelintRunner with customSyntax', () => {
	test('should work properly if customSyntax is defined', async () => {
		expect.assertions(1);
		const runner = new StylelintRunner();
		const result = await runner.lintDocument(
			createDocument('test.css', 'css', 'a\n  color:#ffffff'),
			{
				config: { rules: { 'color-hex-length': 'short' } },
				customSyntax: 'postcss-sass',
			},
		);

		expect(result).toMatchSnapshot();
	});

	test('auto-fix should work properly if customSyntax is defined', async () => {
		expect.assertions(1);
		const runner = new StylelintRunner();

		try {
			const result = await runner.lintDocument(
				createDocument('test.css', 'css', 'a\n  color:#ffffff'),
				{
					config: { rules: { 'color-hex-length': 'short' } },
					customSyntax: 'postcss-sass',
					fix: true,
				},
			);

			expect(result).toMatchSnapshot();
		} catch (e) {
			console.error(e);
		}
	});
});

describe('StylelintRunner with reportDescriptionlessDisables', () => {
	test('should work properly if reportDescriptionlessDisables is true', async () => {
		expect.assertions(1);
		const runner = new StylelintRunner();
		const result = await runner.lintDocument(
			createDocument(
				'test.css',
				'css',
				`
.baz {
    /* stylelint-disable-next-line color-no-invalid-hex */
  color: #y3;
}
/* stylelint-disable color-no-invalid-hex */
.baz {
  color: #y3;
}
/* stylelint-enable color-no-invalid-hex */
.baz {
  color: #y3; /* stylelint-disable-line color-no-invalid-hex */
}

.baz {
    /* stylelint-disable-next-line color-no-invalid-hex -- with a description */
  color: #y3;
}
/* stylelint-disable color-no-invalid-hex -- with a description */
.baz {
  color: #y3;
}
/* stylelint-enable color-no-invalid-hex */
.baz {
  color: #y3; /* stylelint-disable-line color-no-invalid-hex -- with a description */
}
`,
			),
			{
				config: { rules: { 'color-no-invalid-hex': [true] } },
				reportDescriptionlessDisables: true,
			},
		);

		expect(result).toMatchSnapshot();
	});
});

describe('StylelintRunner with reportNeedlessDisables', () => {
	test('should work properly if reportNeedlessDisables is true', async () => {
		expect.assertions(1);
		const runner = new StylelintRunner();
		const result = await runner.lintDocument(
			createDocument(
				'test.css',
				'css',
				`
.foo {
    background-color: #y3; /* stylelint-disable-next-line color-no-invalid-hex */
    color: red;
}

/* stylelint-disable color-no-invalid-hex */
.bar {
    color: red;
}
/* stylelint-enable color-no-invalid-hex */

.baz {
    color: red; /* stylelint-disable-line color-no-invalid-hex */
}

/* stylelint-disable color-no-invalid-hex */
.bar {
    color: red;
}
`,
			),
			{
				config: { rules: { 'color-no-invalid-hex': [true] } },
				reportNeedlessDisables: true,
			},
		);

		expect(result).toMatchSnapshot();
	});
});

describe('StylelintRunner with reportInvalidScopeDisables', () => {
	test('should work properly if reportInvalidScopeDisables is true', async () => {
		expect.assertions(1);
		const runner = new StylelintRunner();
		const result = await runner.lintDocument(
			createDocument(
				'test.css',
				'css',
				`
/* stylelint-disable-next-line foo */

/* stylelint-disable-line foo */

/* stylelint-disable foo */
/* stylelint-enable foo */

/* stylelint-disable-next-line color-no-invalid-hex */

/* stylelint-disable-line color-no-invalid-hex */

/* stylelint-disable color-no-invalid-hex */
/* stylelint-enable color-no-invalid-hex */
`,
			),
			{
				config: { rules: { 'color-no-invalid-hex': [true] } },
				reportInvalidScopeDisables: true,
			},
		);

		expect(result).toMatchSnapshot();
	});
});

describe('StylelintRunner with stylelintPath', () => {
	test('should work properly if stylelintPath is defined', async () => {
		expect.assertions(1);
		const runner = new StylelintRunner();
		const result = await runner.lintDocument(
			createDocument('test.css', 'css', 'a{\n  color:#y3}'),
			{
				config: { rules: { 'color-no-invalid-hex': [true] } },
			},
			{
				stylelintPath: resolve(__dirname, '../../../node_modules/stylelint'),
			},
		);

		expect(result).toMatchSnapshot();
	});

	test('should work properly if custom path is defined in stylelintPath', async () => {
		expect.assertions(1);
		const runner = new StylelintRunner();
		const result = await runner.lintDocument(
			createDocument('test.css', 'css', 'a{\n  color:#y3}'),
			{
				config: { rules: { 'color-no-invalid-hex': [true] } },
			},
			{
				stylelintPath: require.resolve('./fake-stylelint'),
			},
		);

		expect(result).toMatchSnapshot();
	});
});
