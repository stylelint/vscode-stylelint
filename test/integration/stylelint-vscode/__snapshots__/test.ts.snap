// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StylelintRunner should be resolved with diagnostics when it lints CSS successfully 1`] = `
[
  {
    "code": "color-no-invalid-hex",
    "codeDescription": {
      "href": "https://stylelint.io/user-guide/rules/color-no-invalid-hex",
    },
    "message": "Unexpected invalid hex color "#y3" (color-no-invalid-hex)",
    "range": {
      "end": {
        "character": 18,
        "line": 0,
      },
      "start": {
        "character": 15,
        "line": 0,
      },
    },
    "severity": 1,
    "source": "Stylelint",
  },
  {
    "code": "selector-type-no-unknown",
    "codeDescription": {
      "href": "https://stylelint.io/user-guide/rules/selector-type-no-unknown",
    },
    "message": "Unexpected unknown type selector "foo" (selector-type-no-unknown)",
    "range": {
      "end": {
        "character": 5,
        "line": 0,
      },
      "start": {
        "character": 2,
        "line": 0,
      },
    },
    "severity": 1,
    "source": "Stylelint",
  },
]
`;

exports[`StylelintRunner should be resolved with diagnostics when the rules include unknown rules 1`] = `
[
  {
    "code": "this-rule-does-not-exist",
    "message": "Unknown rule this-rule-does-not-exist.",
    "range": {
      "end": {
        "character": 4,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "severity": 1,
    "source": "Stylelint",
  },
  {
    "code": "this-rule-also-does-not-exist",
    "message": "Unknown rule this-rule-also-does-not-exist.",
    "range": {
      "end": {
        "character": 4,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "severity": 1,
    "source": "Stylelint",
  },
]
`;

exports[`StylelintRunner should be resolved with one diagnostic when the CSS is broken 1`] = `
[
  {
    "code": "CssSyntaxError",
    "message": "Unclosed block (CssSyntaxError)",
    "range": {
      "end": {
        "character": 11,
        "line": 0,
      },
      "start": {
        "character": 10,
        "line": 0,
      },
    },
    "severity": 1,
    "source": "Stylelint",
  },
]
`;

exports[`StylelintRunner should check CSS syntax even if no configuration is provided 1`] = `
[
  {
    "code": "CssSyntaxError",
    "message": "Unclosed bracket (CssSyntaxError)",
    "range": {
      "end": {
        "character": 13,
        "line": 0,
      },
      "start": {
        "character": 12,
        "line": 0,
      },
    },
    "severity": 1,
    "source": "Stylelint",
  },
]
`;

exports[`StylelintRunner should check CSS syntax even if no rule is provided 1`] = `
[
  {
    "code": "CssSyntaxError",
    "message": "At-rule without name (CssSyntaxError)",
    "range": {
      "end": {
        "character": 1,
        "line": 0,
      },
      "start": {
        "character": 0,
        "line": 0,
      },
    },
    "severity": 1,
    "source": "Stylelint",
  },
]
`;

exports[`StylelintRunner should reject with a reason when it takes incorrect options 1`] = `
"Invalid option value "foo" for rule "color-hex-alpha"
Invalid option name "bar" for rule "selector-type-no-unknown""
`;

exports[`StylelintRunner should support CSS-in-JS with customSyntax 1`] = `
[
  {
    "code": "font-weight-notation",
    "codeDescription": {
      "href": "https://stylelint.io/user-guide/rules/font-weight-notation",
    },
    "message": "Expected "bold" to be "700" (font-weight-notation)",
    "range": {
      "end": {
        "character": 33,
        "line": 2,
      },
      "start": {
        "character": 29,
        "line": 2,
      },
    },
    "severity": 1,
    "source": "Stylelint",
  },
  {
    "code": "font-weight-notation",
    "codeDescription": {
      "href": "https://stylelint.io/user-guide/rules/font-weight-notation",
    },
    "message": "Expected "normal" to be "400" (font-weight-notation)",
    "range": {
      "end": {
        "character": 12,
        "line": 4,
      },
      "start": {
        "character": 6,
        "line": 4,
      },
    },
    "severity": 1,
    "source": "Stylelint",
  },
]
`;

exports[`StylelintRunner with a configuration file should adhere to configuration file settings 1`] = `
[
  {
    "code": "length-zero-no-unit",
    "codeDescription": {
      "href": "https://stylelint.io/user-guide/rules/length-zero-no-unit",
    },
    "message": "Unexpected unit (length-zero-no-unit)",
    "range": {
      "end": {
        "character": 12,
        "line": 3,
      },
      "start": {
        "character": 10,
        "line": 3,
      },
    },
    "severity": 1,
    "source": "Stylelint",
  },
]
`;

exports[`StylelintRunner with auto-fix auto-fix should work if there are errors that cannot be auto-fixed 1`] = `
{
  "diagnostics": [
    {
      "code": "selector-type-no-unknown",
      "codeDescription": {
        "href": "https://stylelint.io/user-guide/rules/selector-type-no-unknown",
      },
      "message": "Unexpected unknown type selector "unknown" (selector-type-no-unknown)",
      "range": {
        "end": {
          "character": 7,
          "line": 1,
        },
        "start": {
          "character": 0,
          "line": 1,
        },
      },
      "severity": 1,
      "source": "Stylelint",
    },
  ],
  "output": "
unknown {
    color: #fff;
    background-color: #fffa;
}
",
}
`;

exports[`StylelintRunner with auto-fix auto-fix should work if there is syntax errors in css 1`] = `
"
.a {
    color: #ffffff;
    background-color: #ffffffaa
}
"
`;

exports[`StylelintRunner with auto-fix auto-fix should work properly if configs are defined 1`] = `
"a
{
color:#fff;
}"
`;

exports[`StylelintRunner with customSyntax auto-fix should work properly if customSyntax is defined 1`] = `
{
  "diagnostics": [],
  "output": "a
  color:#fff",
}
`;

exports[`StylelintRunner with customSyntax should work properly if customSyntax is defined 1`] = `
{
  "diagnostics": [
    {
      "code": "color-hex-length",
      "codeDescription": {
        "href": "https://stylelint.io/user-guide/rules/color-hex-length",
      },
      "message": "Expected "#ffffff" to be "#fff" (color-hex-length)",
      "range": {
        "end": {
          "character": 15,
          "line": 1,
        },
        "start": {
          "character": 8,
          "line": 1,
        },
      },
      "severity": 1,
      "source": "Stylelint",
    },
  ],
}
`;

exports[`StylelintRunner with reportDescriptionlessDisables should work properly if reportDescriptionlessDisables is true 1`] = `
{
  "diagnostics": [
    {
      "code": "--report-descriptionless-disables",
      "message": "Disable for "color-no-invalid-hex" is missing a description",
      "range": {
        "end": {
          "character": 57,
          "line": 2,
        },
        "start": {
          "character": 4,
          "line": 2,
        },
      },
      "severity": 1,
      "source": "Stylelint",
    },
    {
      "code": "--report-descriptionless-disables",
      "message": "Disable for "color-no-invalid-hex" is missing a description",
      "range": {
        "end": {
          "character": 43,
          "line": 5,
        },
        "start": {
          "character": 0,
          "line": 5,
        },
      },
      "severity": 1,
      "source": "Stylelint",
    },
    {
      "code": "--report-descriptionless-disables",
      "message": "Disable for "color-no-invalid-hex" is missing a description",
      "range": {
        "end": {
          "character": 62,
          "line": 11,
        },
        "start": {
          "character": 14,
          "line": 11,
        },
      },
      "severity": 1,
      "source": "Stylelint",
    },
  ],
}
`;

exports[`StylelintRunner with reportInvalidScopeDisables should work properly if reportInvalidScopeDisables is true 1`] = `
{
  "diagnostics": [
    {
      "code": "--report-invalid-scope-disables",
      "message": "Rule "foo" isn't enabled",
      "range": {
        "end": {
          "character": 36,
          "line": 1,
        },
        "start": {
          "character": 0,
          "line": 1,
        },
      },
      "severity": 1,
      "source": "Stylelint",
    },
    {
      "code": "--report-invalid-scope-disables",
      "message": "Rule "foo" isn't enabled",
      "range": {
        "end": {
          "character": 31,
          "line": 3,
        },
        "start": {
          "character": 0,
          "line": 3,
        },
      },
      "severity": 1,
      "source": "Stylelint",
    },
    {
      "code": "--report-invalid-scope-disables",
      "message": "Rule "foo" isn't enabled",
      "range": {
        "end": {
          "character": 26,
          "line": 5,
        },
        "start": {
          "character": 0,
          "line": 5,
        },
      },
      "severity": 1,
      "source": "Stylelint",
    },
  ],
}
`;

exports[`StylelintRunner with reportNeedlessDisables should work properly if reportNeedlessDisables is true 1`] = `
{
  "diagnostics": [
    {
      "code": "color-no-invalid-hex",
      "codeDescription": {
        "href": "https://stylelint.io/user-guide/rules/color-no-invalid-hex",
      },
      "message": "Unexpected invalid hex color "#y3" (color-no-invalid-hex)",
      "range": {
        "end": {
          "character": 25,
          "line": 2,
        },
        "start": {
          "character": 22,
          "line": 2,
        },
      },
      "severity": 1,
      "source": "Stylelint",
    },
    {
      "code": "--report-needless-disables",
      "message": "Needless disable for "color-no-invalid-hex"",
      "range": {
        "end": {
          "character": 80,
          "line": 2,
        },
        "start": {
          "character": 27,
          "line": 2,
        },
      },
      "severity": 1,
      "source": "Stylelint",
    },
    {
      "code": "--report-needless-disables",
      "message": "Needless disable for "color-no-invalid-hex"",
      "range": {
        "end": {
          "character": 43,
          "line": 6,
        },
        "start": {
          "character": 0,
          "line": 6,
        },
      },
      "severity": 1,
      "source": "Stylelint",
    },
    {
      "code": "--report-needless-disables",
      "message": "Needless disable for "color-no-invalid-hex"",
      "range": {
        "end": {
          "character": 64,
          "line": 13,
        },
        "start": {
          "character": 16,
          "line": 13,
        },
      },
      "severity": 1,
      "source": "Stylelint",
    },
    {
      "code": "--report-needless-disables",
      "message": "Needless disable for "color-no-invalid-hex"",
      "range": {
        "end": {
          "character": 43,
          "line": 16,
        },
        "start": {
          "character": 0,
          "line": 16,
        },
      },
      "severity": 1,
      "source": "Stylelint",
    },
  ],
}
`;

exports[`StylelintRunner with stylelintPath should work properly if custom path is defined in stylelintPath 1`] = `
{
  "diagnostics": [
    {
      "code": "fake",
      "message": "Fake result",
      "range": {
        "end": {
          "character": 1,
          "line": 0,
        },
        "start": {
          "character": 0,
          "line": 0,
        },
      },
      "severity": 1,
      "source": "Stylelint",
    },
  ],
}
`;

exports[`StylelintRunner with stylelintPath should work properly if stylelintPath is defined 1`] = `
{
  "diagnostics": [
    {
      "code": "color-no-invalid-hex",
      "codeDescription": {
        "href": "https://stylelint.io/user-guide/rules/color-no-invalid-hex",
      },
      "message": "Unexpected invalid hex color "#y3" (color-no-invalid-hex)",
      "range": {
        "end": {
          "character": 11,
          "line": 1,
        },
        "start": {
          "character": 8,
          "line": 1,
        },
      },
      "severity": 1,
      "source": "Stylelint",
    },
  ],
}
`;
